lib/skate/migrate.ex

```diff
@@ -1,5 +1,53 @@
 defmodule Skate.Migrate do
-  def up() do
-    Ecto.Migrator.run(Skate.Repo, :up, all: true)
+  @moduledoc """
+  GenServer which runs on startup to run Ecto migrations. All migrations stored in the "migrations" directory are run during init. Migrations stored in the "async_migrations" directory will be run after the regular migrations complete and will only log a warning on failure.
+  """
+  use GenServer, restart: :transient
+  require Logger
+
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @impl GenServer
+  def init(opts) do
+    Logger.info("#{__MODULE__} synchronous migrations starting")
+    Keyword.get(opts, :run_sync_migrations, default_migrator_fn("migrations")).()
+
+    Logger.info("#{__MODULE__} synchronous migrations finished")
+    {:ok, opts, {:continue, :async_migrations}}
+  end
+
+  @impl GenServer
+  def handle_continue(:async_migrations, opts) do
+    Logger.info("#{__MODULE__} async migrations starting")
+
+    try do
+      Keyword.get(
+        opts,
+        :run_async_migrations,
+        default_migrator_fn("async_migrations")
+      ).()
+
+      Logger.info("#{__MODULE__} async migrations finished")
+    rescue
+      e ->
+        Logger.warn("#{__MODULE__} async migrations failed. error=#{inspect(e)}")
+        :ok
+    end
+
+    {:stop, :normal, opts}
+  end
+
+  @spec default_migrator_fn(String.t()) :: function()
+  defp default_migrator_fn(migration_directory) do
```

Comment:
**suggestion (non-blocking):** I think it would be slightly clearer to call this `migrator` or `migrate` or something like that, and then call the `opts` keyword `:migrator_fn`.

Reply:
Oh, I see, this actually _returns_ a function. I think it could be refactored to not do that, though, basically:

```
Keyword.get(opts, :run_sync_migrations, default_migrator_fn("migrations")).()
```

could become

```
Keyword.get(opts, :run_sync_migrations, default_migrator_fn).("migrations")
```

so that there's only one level of function call, and the mock test functions could be updated as needed.